src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -178,21 +178,28 @@ internal GeneratorDriverState RunGeneratorsCore(Compilation compilation, Diagnos
                                      : SetGeneratorException(MessageProvider, GeneratorState.Uninitialized, generator, ex, diagnosticsBag, isInit: true);
                 }
 
-                // create the syntax receiver if requested
-                if (generatorState.Info.SyntaxReceiverCreator is object)
+                // create the syntax receivers if needed
+                Debug.Assert(!(generatorState.Info.SyntaxReceiverCreator is object && generatorState.Info.SyntaxContextReceiverCreator is object));
```

Comment:
It seems awkward that throughout this change we have to constantly be validating that we have only one of these set. It adds this complexity throughout the code base. 

Did you consider instead the idea of just making `ISyntaxReceiver` an implementation detail of the registration process? Consider that you could define an adapter class that implements `ISyntaxReceiverWithContext` and just wraps a `ISyntaxReceiver` instance. The registration layer could apply this adapter and the rest of the code base could essentially be ignorant of the fact that we have `ISyntaxReceiver` at all.

Reply:
Yeah I think that might be a better way to go, we still need to 'unwrap' the adaptor if you request it in the generator context, but that seems simple enough. Let me play with that idea a bit.